File: C# Fundamentals/C# OOP Advanced/07.Iterators_Comparators/Ex/P07_EqualityLogic/Person.cs
using System;

public class Person : IComparable<Person>
{
    public string Name { get; set; }

    public int Age { get; set; }

    public Person(string name, int age)
    {
        this.Name = name;
        this.Age = age;
    }

    public override string ToString()
    {
        return $"{this.Name} {this.Age}";
    }

    public int CompareTo(Person other)
    {
        int result = this.Name.CompareTo(other.Name);
        if(result == 0)
        {
            result = this.Age.CompareTo(other.Age);
        }
        return result;
    }
    public override bool Equals(object obj)
    {
        Person secondPerson = obj as Person;

        int result = this.Name.CompareTo(secondPerson.Name);
        if (secondPerson == null || result != 0)
        {
            return false;
        }
        result = this.Age.CompareTo(secondPerson.Age);
        if (result != 0)
        {
            return false;
        }
        return true;
    }
    public override int GetHashCode()
    {
        return this.Name.GetHashCode() ^ this.Age.GetHashCode();
    }
}

File: Java/01.Advanced/01.Advanced/17.IteratorsComparators/Ex/src/t_05_comparing_objects/Person.java
package t_05_comparing_objects;

public class Person implements Comparable<Person>{
    private final String name;
    private final int age;
    private final String town;

    public Person(String name, int age, String town) {
        this.name = name;
        this.age = age;
        this.town = town;
    }

    @Override
    public int compareTo(Person other) {
        int result = this.name.compareTo(other.name);

        if (result == 0) {
            result = Integer.compare(this.age, other.age);
        }

        if (result == 0) {
            return this.town.compareTo(other.town);
        }

        return result;
    }
}
